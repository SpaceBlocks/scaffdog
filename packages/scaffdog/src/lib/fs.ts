import { readFileSync, statSync } from 'fs';
import { mkdir, readFile, writeFile } from 'fs/promises';
import type { GlobbyOptions } from 'globby';
import globby from 'globby';

export type FsMkdirOptions = {
  recursive?: boolean;
};

export type FsGlobOptions = GlobbyOptions;

export type FsLibrary = {
  mkdir: (filepath: string, options: FsMkdirOptions) => Promise<void>;
  readFile: (filepath: string) => Promise<string>;
  readFileSync: (filepath: string) => string;
  glob: (
    patterns: string | readonly string[],
    options: FsGlobOptions,
  ) => Promise<string[]>;
  writeFile: (filepath: string, data: string) => Promise<void>;
  fileExists: (filepath: string) => boolean;
  directoryExists: (filepath: string) => boolean;
};

export const createFsLibrary = (): FsLibrary => {
  const fstat = (filepath: string) => {
    try {
      return statSync(filepath);
    } catch (e) {
      return null;
    }
  };

  return {
    mkdir: async (filepath, options) => {
      await mkdir(filepath, options);
    },

    readFile: async (filepath) => {
      return await readFile(filepath, 'utf8');
    },

    readFileSync: (filepath) => {
      return readFileSync(filepath, 'utf8');
    },

    glob: async (patterns, options) => {
      return await globby(patterns, options);
    },

    writeFile: async (filepath, data) => {
      return await writeFile(filepath, data, 'utf8');
    },

    fileExists: (filepath) => {
      return fstat(filepath)?.isFile() ?? false;
    },

    directoryExists: (filepath) => {
      return fstat(filepath)?.isDirectory() ?? false;
    },
  };
};
