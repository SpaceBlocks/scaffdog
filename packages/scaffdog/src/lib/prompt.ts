/// <reference types="../types/inquirer-autocomplete-prompt" />
import Fuse from 'fuse.js';
import inquirer from 'inquirer';
import inquirerAutocompletePrompt from 'inquirer-autocomplete-prompt';

inquirer.registerPrompt('autocomplete', inquirerAutocompletePrompt);

export type PromptQuestion = inquirer.DistinctQuestion;

export type ConfirmArgs = Omit<PromptQuestion, 'type' | 'name' | 'default'>;

export type AutocompleteArgs = Omit<PromptQuestion, 'type' | 'name'>;

export type PromptLibrary = {
  prompt: <T>(question: PromptQuestion) => Promise<T>;

  confirm: (
    message: string,
    initial: boolean,
    args: ConfirmArgs,
  ) => Promise<boolean>;

  autocomplete: (
    message: string,
    list: string[],
    args: AutocompleteArgs,
  ) => Promise<string>;
};

export const createPromptLibrary = (): PromptLibrary => {
  const prompt: PromptLibrary['prompt'] = async (question) => {
    const { value } = await inquirer.prompt([
      {
        ...question,
        name: 'value',
      },
    ]);

    if (question.when === false && question.default != null) {
      return question.default;
    }

    return value;
  };

  return {
    prompt,

    confirm: async (message, initial, args) => {
      return await prompt({
        ...args,
        type: 'confirm',
        default: initial,
        message,
      });
    },

    autocomplete: async (message, list, args) => {
      const fuse = new Fuse(list, {});

      return await prompt({
        ...args,
        type: 'autocomplete',
        message,
        pageSize: 20,
        source: (_: any, input: string) => {
          return !input
            ? list
            : fuse.search(input || '').map(({ item }) => item);
        },
      } as any);
    },
  };
};
