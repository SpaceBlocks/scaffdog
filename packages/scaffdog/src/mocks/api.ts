import { vi } from 'vitest';
import type { Scaffdog, ScaffdogInitializer } from '../api';
import { createResolvedConfig } from '../lib/config.factory';

export const createScaffdogMock = (
  mocks: Partial<Scaffdog> = {},
): Scaffdog => ({
  path: {
    project: '',
    config: '',
  },
  config: createResolvedConfig(),
  list: vi.fn(),
  generate: vi.fn(),
  ...mocks,
});

export const createScaffdogInitializerMock = (
  mocks: Partial<ScaffdogInitializer> = {},
): ScaffdogInitializer => {
  const scaffdog = createScaffdogMock();

  return {
    createScaffdog: vi.fn().mockReturnValueOnce(scaffdog),
    loadScaffdog: vi.fn().mockResolvedValueOnce(scaffdog),
    ...mocks,
  };
};
