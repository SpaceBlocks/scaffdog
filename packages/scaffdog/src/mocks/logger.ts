import { PassThrough } from 'stream';
import type { ConsolaInstance } from 'consola';
import { LogLevels, createConsola } from 'consola';
import stripAnsi from 'strip-ansi';

export const createLogger = (): {
  logger: ConsolaInstance;
  getStdout: () => string;
  getStderr: () => string;
} => {
  const output = {
    stdout: '',
    stderr: '',
  };

  const stream = {
    stdout: new PassThrough(),
    stderr: new PassThrough(),
  };

  const logger = createConsola({
    fancy: true,
    formatOptions: {
      date: false,
    },
    level: LogLevels.info,
    ...(stream as any),
  });

  stream.stdout.on('data', (chunk) => (output.stdout += chunk));
  stream.stderr.on('data', (chunk) => (output.stderr += chunk));

  return {
    logger,
    getStdout: () => stripAnsi(output.stdout),
    getStderr: () => stripAnsi(output.stderr),
  };
};
