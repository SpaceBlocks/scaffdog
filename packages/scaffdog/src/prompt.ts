/// <reference types="./types/inquirer-autocomplete-prompt" />
import Fuse from 'fuse.js';
import inquirer from 'inquirer';
import inquirerAutocompletePrompt from 'inquirer-autocomplete-prompt';

inquirer.registerPrompt('autocomplete', inquirerAutocompletePrompt);

export const prompt = async <T>(
  question: inquirer.DistinctQuestion,
): Promise<T> => {
  const { value } = await inquirer.prompt<{ value: T }>([
    {
      ...question,
      name: 'value',
    },
  ]);

  if (question.when === false && question.default != null) {
    return question.default;
  }

  return value;
};

export const confirm = async (
  message: string,
  initial: boolean,
  args: Omit<inquirer.DistinctQuestion, 'type' | 'name' | 'default'> = {},
): Promise<boolean> => {
  return await prompt({
    ...args,
    type: 'confirm',
    default: initial,
    message,
  });
};

export const autocomplete = async (
  message: string,
  list: string[],
  args: Omit<inquirer.DistinctQuestion, 'type' | 'name'> = {},
): Promise<string> => {
  const fuse = new Fuse(list, {});

  return await prompt({
    ...args,
    type: 'autocomplete',
    message,
    pageSize: 20,
    source: (_: any, input: string) => {
      return !input ? list : fuse.search(input || '').map(({ item }) => item);
    },
  } as any);
};
