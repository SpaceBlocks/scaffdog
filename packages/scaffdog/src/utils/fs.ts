import fs from 'fs';
import { promisify } from 'util';

export const writeFile = promisify(fs.writeFile);
export const readFile = promisify(fs.readFile);
export const mkdir = promisify(fs.mkdir);

const fstat = (filepath: string) => {
  try {
    return fs.statSync(filepath);
  } catch (e) {
    return null;
  }
};

export const fileExists = (filepath: string): boolean =>
  fstat(filepath)?.isFile() ?? false;

export const directoryExists = (filepath: string): boolean =>
  fstat(filepath)?.isDirectory() ?? false;
