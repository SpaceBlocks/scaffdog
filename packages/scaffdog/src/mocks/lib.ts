import type { Library } from '../lib';
import { createConfigLibraryMock } from '../lib/config.mock';
import { createDocumentLibraryMock } from '../lib/document.mock';
import { createErrorLibraryMock } from '../lib/error.mock';
import { createFsLibraryMock } from '../lib/fs.mock';
import { createPromptLibraryMock } from '../lib/prompt.mock';
import { createQuestionLibraryMock } from '../lib/question.mock';

export const createLibraryMock = (mocks: Partial<Library> = {}): Library => {
  return {
    fs: createFsLibraryMock(),
    error: createErrorLibraryMock(),
    prompt: createPromptLibraryMock(),
    config: createConfigLibraryMock(),
    question: createQuestionLibraryMock(),
    document: createDocumentLibraryMock(),
    ...mocks,
  };
};
