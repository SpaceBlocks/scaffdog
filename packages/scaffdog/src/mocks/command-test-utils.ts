import path from 'path';
import type {
  AnyCommandModule,
  CommandArgs,
  CommandModule,
  CommandOption,
  GlobalFlags,
  GlobalFlagsWith,
  InferredCommandModuleArgs,
  InferredCommandModuleFlags,
} from '../command';
import type { CommandContainer } from '../command-container';
import { createCommandContainer } from '../command-container';
import type { Library } from '../lib';
import { createLibraryMock } from './lib';
import { createLogger } from './logger';

export const cwd = path.resolve(__dirname, '../../');

export type CommandRunnerParams<T extends AnyCommandModule> = {
  args?: Partial<CommandArgs<InferredCommandModuleArgs<T>>>;
  flags?: Partial<GlobalFlagsWith<CommandArgs<InferredCommandModuleFlags<T>>>>;
  lib?: Library;
  container?: CommandContainer;
};

export type CommandRunner<T extends AnyCommandModule> = (
  params: CommandRunnerParams<T>,
) => Promise<{
  code: number;
  stdout: string;
  stderr: string;
}>;

export const createCommandRunner =
  <A extends CommandOption, F extends CommandOption>(
    cmd: CommandModule<A, F>,
    defaults: {
      args: CommandArgs<A>;
      flags: CommandArgs<F>;
    },
  ): CommandRunner<CommandModule<A, F>> =>
  async ({ args, flags, lib, container }) => {
    const { logger, getStdout, getStderr } = createLogger();

    const globalFlags: GlobalFlags = {
      project: '.scaffdog',
      help: false,
      version: false,
      verbose: true,
    };

    const code = await cmd.run({
      cwd,
      pkg: {
        name: 'scaffdog',
        version: '1.0.0',
        description: 'scaffdog is Markdown driven scaffolding tool.',
      },
      size: {
        rows: 120,
        columns: 60,
      },
      logger,
      container: container ?? createCommandContainer([]),
      lib: lib ?? createLibraryMock({}),
      args: {
        ...defaults.args,
        ...(args ?? {}),
      },
      flags: {
        ...defaults.flags,
        ...globalFlags,
        ...(flags ?? {}),
      } as never,
    });

    return {
      code,
      stdout: getStdout(),
      stderr: getStderr(),
    };
  };
