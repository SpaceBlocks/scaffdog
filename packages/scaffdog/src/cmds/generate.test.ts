import path from 'path';
import fs from 'fs';
import { test, describe, expect, afterEach, beforeAll, vi } from 'vitest';
import { cwd, runCommand } from '../mocks/command-test-utils';
import * as prompt from '../prompt';
import cmd from './generate';

const defaults = {
  name: undefined,
  'dry-run': false,
  force: false,
};

const clear = () => {
  try {
    fs.rmSync(path.resolve(cwd, 'tmp'), {
      recursive: true,
    });
  } catch (e) {}
};

const mount = (files: Record<string, { path: string; content: string }>) => {
  for (const entry of Object.values(files)) {
    const filepath = path.resolve(cwd, entry.path);
    const dirname = path.dirname(filepath);

    fs.mkdirSync(dirname, { recursive: true });
    fs.writeFileSync(filepath, entry.content, 'utf8');
  }
};

beforeAll(() => {
  clear();
});

afterEach(() => {
  clear();
  vi.clearAllMocks();
});

describe('prompt', () => {
  test('name', async () => {
    vi.spyOn(prompt, 'prompt').mockResolvedValueOnce('a'); // name

    const { code, stdout, stderr } = await runCommand(cmd, {
      ...defaults,
    });

    expect(stderr).toBe('');
    expect(stdout).toMatchSnapshot();
    expect(code).toBe(0);
    expect(
      fs.readFileSync(path.resolve(cwd, 'tmp/nest/dump.txt'), 'utf8'),
    ).toMatchSnapshot();
  });

  test('overwrite files', async () => {
    vi.spyOn(prompt, 'confirm')
      .mockResolvedValueOnce(true) // dump.txt
      .mockResolvedValueOnce(false); // generate.txt (skipped)

    const file = {
      dump: {
        path: path.resolve(cwd, 'tmp/nest/dump.txt'),
        content: 'dump',
      },
      generate: {
        path: path.resolve(cwd, 'tmp/generate.txt'),
        content: 'generate',
      },
    };

    mount(file);

    const { code, stdout, stderr } = await runCommand(cmd, {
      ...defaults,
      name: 'a',
    });

    expect(stderr).toBe('');
    expect(stdout).toMatchSnapshot();
    expect(code).toBe(0);

    expect(fs.readFileSync(file.dump.path, 'utf8')).not.toBe(file.dump.content);
    expect(fs.readFileSync(file.generate.path, 'utf8')).toBe(
      file.generate.content,
    );
  });

  test('has magic and question', async () => {
    fs.mkdirSync(path.resolve(cwd, 'tmp/root'), { recursive: true });

    vi.spyOn(prompt, 'autocomplete').mockResolvedValueOnce('root');
    vi.spyOn(prompt, 'prompt')
      .mockResolvedValueOnce('value')
      .mockResolvedValueOnce('B');

    const { code, stdout, stderr } = await runCommand(cmd, {
      ...defaults,
      name: 'b',
    });

    expect(stderr).toBe('');
    expect(stdout).toMatchSnapshot();
    expect(code).toBe(0);
  });

  test('all question types', async () => {
    vi.spyOn(prompt, 'prompt')
      .mockResolvedValueOnce('shorthand') // shorthand
      .mockResolvedValueOnce('input') // input
      .mockResolvedValueOnce('input_with_initial') // input_with_initial
      .mockResolvedValueOnce('input_if') // input_if
      .mockResolvedValueOnce(true) // bool
      .mockResolvedValueOnce(true) // bool_with_true
      .mockResolvedValueOnce(false) // bool_with_false
      .mockResolvedValueOnce(true) // bool_if
      .mockResolvedValueOnce('C') // list
      .mockResolvedValueOnce('B') // list_with_initial
      .mockResolvedValueOnce('A') // list_if
      .mockResolvedValueOnce(['A', 'C']) // checkbox
      .mockResolvedValueOnce(['B', 'C']) // checkbox_with_initial
      .mockResolvedValueOnce(['A']); // checkbox_if

    const { code, stdout, stderr } = await runCommand(cmd, {
      ...defaults,
      name: 'question',
    });

    expect(stderr).toBe('');
    expect(stdout).toMatchSnapshot();
    expect(code).toBe(0);

    expect(
      fs.readFileSync(path.resolve(cwd, 'tmp/result.txt'), 'utf8'),
    ).toMatchSnapshot();
  });
});

describe('options', () => {
  test('options', async () => {
    const { code, stdout, stderr } = await runCommand(cmd, {
      ...defaults,
      name: 'a',
    });

    expect(code).toBe(0);
    expect(stderr).toBe('');
    expect(stdout).toMatchSnapshot();
  });

  test('force overwrite', async () => {
    vi.spyOn(prompt, 'confirm').mockRejectedValueOnce(new Error('unexpected'));

    const file = {
      dump: {
        path: path.resolve(cwd, 'tmp/nest/dump.txt'),
        content: 'dump',
      },
      generate: {
        path: path.resolve(cwd, 'tmp/generate.txt'),
        content: 'generate',
      },
    };

    mount(file);

    const { code, stdout, stderr } = await runCommand(cmd, {
      ...defaults,
      name: 'a',
      force: true,
    });

    expect(stderr).toBe('');
    expect(stdout).toMatchSnapshot();
    expect(code).toBe(0);

    expect(fs.readFileSync(file.dump.path, 'utf8')).not.toBe(file.dump.content);
    expect(fs.readFileSync(file.generate.path, 'utf8')).not.toBe(
      file.generate.content,
    );
  });

  test('dry run', async () => {
    const { code, stdout, stderr } = await runCommand(cmd, {
      ...defaults,
      name: 'a',
      'dry-run': true,
    });

    expect(stderr).toBe('');
    expect(stdout).toMatchSnapshot();
    expect(code).toBe(0);
  });

  test('not found', async () => {
    const { code, stdout, stderr } = await runCommand(cmd, {
      ...defaults,
      name: 'not-found',
    });

    expect(stderr).toMatchSnapshot();
    expect(stdout).toBe('');
    expect(code).toBe(1);
  });
});
