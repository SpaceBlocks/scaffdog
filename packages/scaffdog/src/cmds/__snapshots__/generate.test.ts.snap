// Vitest Snapshot v1

exports[`options > dry run 1`] = `
"ℹ Output destination directory: \\".\\"

───┬────────────────────────────────────────────────────────
 ✔ │ File: tmp/nest/dump.txt
───┼────────────────────────────────────────────────────────
 1 │ a
 2 │ tmp/nest/dump.txt
 3 │ dump
 4 │ dump.txt
 5 │ .txt
 6 │ tmp/nest
 7 │ dump.txt
 8 │ 
 9 │ ../../snapshots/src

───┬────────────────────────────────────────────────────────
 ✔ │ File: tmp/generate.txt
───┼────────────────────────────────────────────────────────
 1 │ foo

🐶 Generated 2 files!

     ✔ tmp/nest/dump.txt
     ✔ tmp/generate.txt
"
`;

exports[`options > force overwrite 1`] = `
"ℹ Output destination directory: \\".\\"

🐶 Generated 2 files!

     ✔ tmp/nest/dump.txt
     ✔ tmp/generate.txt
"
`;

exports[`options > not found 1`] = `
"
 ERROR  Document \\"not-found\\" not found.

"
`;

exports[`options > options 1`] = `
"ℹ Output destination directory: \\".\\"

🐶 Generated 2 files!

     ✔ tmp/nest/dump.txt
     ✔ tmp/generate.txt
"
`;

exports[`prompt > all question types 1`] = `
"ℹ Output destination directory: \\"tmp\\"

🐶 Generated 1 file!

     ✔ tmp/result.txt
"
`;

exports[`prompt > all question types 2`] = `
"shorthand: shorthand
input: input
input_with_initial: input_with_initial
input_if: input_if
input_if_with_initial: default
input_if_without_initial: 
bool: true
bool_with_true: true
bool_with_false: false
bool_if: true
bool_if_with_initial: true
bool_if_without_initial: false
list: C
list_with_initial: B
list_if: A
list_if_with_initial: B
list_if_without_initial: 
checkbox: A,C
checkbox_with_initial: B,C
checkbox_if_with_initial: B,C
checkbox_if_without_initial: "
`;

exports[`prompt > has magic and question 1`] = `
"ℹ Output destination directory: \\"tmp/root\\"

🐶 Generated 1 file!

     ✔ tmp/root/b.txt
"
`;

exports[`prompt > name 1`] = `
"ℹ Output destination directory: \\".\\"

🐶 Generated 2 files!

     ✔ tmp/nest/dump.txt
     ✔ tmp/generate.txt
"
`;

exports[`prompt > name 2`] = `
"a
tmp/nest/dump.txt
dump
dump.txt
.txt
tmp/nest
dump.txt

../../snapshots/src"
`;

exports[`prompt > overwrite files 1`] = `
"ℹ Output destination directory: \\".\\"

🐶 Generated 1 file! (1 skipped)

     ✔ tmp/nest/dump.txt
     ⚠ tmp/generate.txt (skipped)
"
`;

exports[`prompt > variables section 1`] = `
"ℹ Output destination directory: \\".\\"

🐶 Generated 1 file!

     ✔ tmp/index.txt
"
`;

exports[`prompt > variables section 2`] = `"foo: success"`;
